templates/model/decode/decode.go

```diff
@@ -107,16 +109,28 @@ func Decode(r io.Reader, filename, requireKind string) (model.ValidatorUpgrader,
 		return nil, "", fmt.Errorf("file %s has kind %q, but %q is required", filename, cf.Kind, requireKind)
 	}
 
-	vu, err := modelForKind(filename, apiVersion, cf.Kind)
-	if err != nil {
-		return nil, "", err
+	vu, err := decodeFromVersionKind(filename, apiVersion, cf.Kind, buf)
+	if err == nil {
+		return vu, apiVersion, nil
 	}
 
-	if err := yaml.Unmarshal(buf, vu); err != nil {
-		return nil, "", fmt.Errorf("error parsing YAML file %s: %w", filename, err)
+	// Parsing or validation failed. We'll try to detect a common user error
+	// that might have caused this problem and print a helpful message. The user
+	// might be trying to use a new feature, but the api_version declared in
+	// their YAML file doesn't support that new feature. To detect this, we'll
+	// speculatively try to parse the YAML file with a newer api_version and see
+	// if that version would have been valid. If so, we inform the user that
+	// they should change the api_version field in their YAML file.
+	attemptAPIVersion := apiVersions[len(apiVersions)-1].apiVersion
+	if attemptAPIVersion == apiVersion {
+		return nil, "", err // api_version upgrade isn't possible, they're already on the latest.
+	}
+	if _, attemptErr := decodeFromVersionKind(filename, attemptAPIVersion, cf.Kind, buf); attemptErr == nil {
+		return nil, "", fmt.Errorf("file %s sets api_version %q but does not parse and validate successfully under that version. However, it will be valid if you change the api_version to %q. The error was: %w",
```

Comment:
```suggestion
		return nil, "", fmt.Errorf("file %s sets api_version %q but does not parse and validate successfully under that version. However, it will be valid if you upgrade the api_version to %q. The error was: %w",
```

Reply:
Just so I don't lose my readability approval, I'll merge as-is and send another one for this change.